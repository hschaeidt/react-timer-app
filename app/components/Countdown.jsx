import React from 'react';
import Clock from './Clock';
import CountdownForm from './CountdownForm';

export const STATUS = {
  STARTED: 'started',
  STOPPED: 'stopped',
};

export default class Countdown extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.startTimer = this.startTimer.bind(this);
    this.handleSetCountdown = this.handleSetCountdown.bind(this);
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.state.status !== prevState.status) {
      switch (this.state.status) {
        case STATUS.STARTED:
          this.startTimer();
          break;
        case STATUS.STOPPED:
          clearInterval(this.timer);
          break;
      }
    }
  }

  startTimer() {
    this.timer = setInterval(() => {
      const secondsRemaining = this.state.count - 1;
      this.setState({
        count: secondsRemaining,
      });

      if (secondsRemaining === 0) {
        this.setState({
          status: STATUS.STOPPED,
        });
      }
    }, 1000);
  }

  handleSetCountdown(seconds) {
    this.setState({
      count: seconds,
      status: STATUS.STARTED,
    });
  }

  render() {
    const { count } = this.state;

    return (
      <div>
        <Clock totalSeconds={count} />
        <CountdownForm onSetCountdown={this.handleSetCountdown} />
      </div>
    );
  }
}
