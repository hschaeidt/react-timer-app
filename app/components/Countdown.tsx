import * as React from 'react';
import Clock from './Clock';
import CountdownForm from './CountdownForm';
import Controls from './Controls';

export enum STATUS {
  STARTED,
  STOPPED,
  PAUSED,
}

interface CountdownProps extends React.Props<any> {}

interface CountdownState {
  count: number,
  status: STATUS,
}

export default class Countdown extends React.Component<CountdownProps, CountdownState> {
  private timer: number;

  constructor(props: CountdownProps) {
    super(props);

    this.state = {
      count: 0,
      status: STATUS.STOPPED,
    };

    this.startTimer = this.startTimer.bind(this);
    this.stopTimer = this.stopTimer.bind(this);
    this.handleSetCountdown = this.handleSetCountdown.bind(this);
    this.handleStatusChange = this.handleStatusChange.bind(this);
  }

  componentDidUpdate(prevProps: CountdownProps, prevState: CountdownState): void {
    if (this.state.status !== prevState.status) {
      switch (this.state.status) {
        case STATUS.STARTED:
          this.startTimer();
          break;
        default:
          this.stopTimer();
          break;
      }
    }
  }

  componentWillUnmount(): void {
    this.stopTimer();
  }

  startTimer(): void {
    this.timer = setInterval(() => {
      const secondsRemaining = this.state.count - 1;
      this.setState({
        count: secondsRemaining,
      });

      if (secondsRemaining === 0) {
        this.handleStatusChange(STATUS.STOPPED);
      }
    }, 1000);
  }

  stopTimer(): void {
    clearInterval(this.timer);
    this.timer = undefined;
  }

  handleSetCountdown(seconds: number) {
    if (seconds > 0) {
      this.setState({
        count: seconds,
        status: STATUS.STARTED,
      });
    }
  }

  handleStatusChange(status: STATUS) {
    const state: CountdownState = { 
      status,
      count: this.state.count,
    };

    if (status === STATUS.STOPPED) {
      state.count = 0;
    }

    this.setState(state);
  }

  render() {
    const { count, status } = this.state;

    const renderControlArea = () => {
      if (status !== STATUS.STOPPED) {
        return <Controls countdownStatus={status} onStatusChange={this.handleStatusChange} />;
      }

      return <CountdownForm onSetCountdown={this.handleSetCountdown} />;
    };

    return (
      <div>
        <h1 className="page-title">Countdown App</h1>
        <Clock totalSeconds={count} />
        {renderControlArea()}
      </div>
    );
  }
}
