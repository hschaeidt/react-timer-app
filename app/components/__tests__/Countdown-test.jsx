import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import Countdown, { STATUS } from '../Countdown';

jest.useFakeTimers();

function setup() {
  const wrapper = shallow(<Countdown />);
  const instance = wrapper.instance();
  const tree = renderer.create(<Countdown />);

  return {
    wrapper,
    instance,
    tree,
  };
}

describe('<Countdown />', () => {
  it('renders correctly', () => {
    const { tree } = setup();
    expect(tree).toMatchSnapshot();
  });

  describe('handleSetCountdown', () => {
    it('should set state to started and counting down', () => {
      const { instance } = setup();
      instance.handleSetCountdown(10);

      expect(instance.state.count).toBe(10);
      expect(instance.state.status).toBe(STATUS.STARTED);

      jest.runOnlyPendingTimers();

      expect(instance.state.count).toBe(9);
    });

    it('should never set count less than zero', () => {
      const { instance } = setup();
      jest.runAllTimers();
      expect(instance.timer).not.toBeDefined();
      expect(instance.state.count).toBe(0);

      instance.handleSetCountdown(0);
      expect(instance.timer).not.toBeDefined();
      jest.runAllTimers();
      expect(instance.state.count).toBe(0);

      instance.handleSetCountdown(10);
      expect(instance.timer).toBeDefined();
      jest.runAllTimers();
      expect(instance.timer).not.toBeDefined();
      expect(instance.state.count).toBe(0);
    });

    it('should pause countdown on paused status', () => {
      const { instance } = setup();
      instance.handleSetCountdown(3);
      instance.handleStatusChange(STATUS.PAUSED);

      jest.runAllTimers();

      expect(instance.state.count).toBe(3);
    });

    it('should reset countdown on stopped status', () => {
      const { instance } = setup();
      instance.handleSetCountdown(10);
      instance.handleStatusChange(STATUS.STOPPED);

      jest.runAllTimers();

      expect(instance.state.count).toBe(0);
      expect(instance.timer).not.toBeDefined();
    });

    it('should clear the timer on unmount', () => {
      const { wrapper, instance } = setup();

      expect(instance.timer).not.toBeDefined();
      instance.handleSetCountdown(10);
      expect(instance.timer).toBeDefined();

      wrapper.unmount();
      expect(instance.timer).not.toBeDefined();
    });
  });
});
