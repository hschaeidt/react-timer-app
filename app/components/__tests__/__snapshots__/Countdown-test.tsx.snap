exports[`<Countdown /> renders correctly 1`] = `
<div>
  <h1
    className="page-title">
    Countdown App
  </h1>
  <div
    className="clock">
    <span
      className="clock-text">
      00:00
    </span>
  </div>
  <form
    className="countdown-form"
    onSubmit={[Function]}>
    <input
      onChange={[Function]}
      placeholder="Enter time in seconds"
      type="text"
      value="" />
    <button
      className="button expanded">
      Start
    </button>
  </form>
</div>
`;
